src/validate.js

```diff
@@ -1,169 +1,180 @@
+      debug('Attempting to set constants by file: %s', fullpath);
+      try {
+        this.cfg.constants = yaml.safeLoad(fs.readFileSync(fullpath, 'utf-8'));
+      } catch (err) {
+        if (err.code == 'ENOENT') {
+          debug('Constants file does not exist, setting constants to {}');
+          this.cfg.constants = {};
+        } else {
+          throw err;
+        }
       }
     }
-  }
 
-  let walkErr;
-  walk.walkSync(path.resolve(cfg.folder), {listeners: {file: (root, stats) => {
-    try {
-      let name = path.relative(cfg.folder, path.join(root, stats.name));
-
-      let json = null;
-      let data = fs.readFileSync(path.join(cfg.folder, name), 'utf-8');
-      if (/\.ya?ml$/.test(name) && name !== 'constants.yml') {
-        json = yaml.safeLoad(data);
-      } else if (/\.json$/.test(name)) {
-        json = JSON.parse(data);
-      } else {
-        debug('Ignoring file %s', name);
-        return;
-      }
+    let walkErr;
+    walk.walkSync(path.resolve(this.cfg.folder), {listeners: {file: (root, stats) => {
+      try {
+        let name = path.relative(this.cfg.folder, path.join(root, stats.name));
+
+        let json = null;
+        const data = fs.readFileSync(path.join(this.cfg.folder, name), 'utf-8');
+        if (/\.ya?ml$/.test(name) && name !== 'constants.yml') {
+          json = yaml.safeLoad(data);
+        } else if (/\.json$/.test(name)) {
+          json = JSON.parse(data);
+        } else {
+          debug('Ignoring file %s', name);
+          return;
+        }
 
-      let schema = render(json, cfg.constants);
+        const jsonName = name.replace(/\.ya?ml$/, '.json');
+        this.rawSchemas[jsonName] = json;
 
-      if (schema.id) {
-        debug('Schema incorrectly attempts to set own id: %s', name);
-        throw new Error('Schema ' + path.join(root, name) + ' attempts to set own id!');
-      }
-      let jsonName = name.replace(/\.ya?ml$/, '.json');
-      schema.id = libUrls.schema(cfg.rootUrl, cfg.serviceName, jsonName + '#');
-
-      ajv.addSchema(schema);
-      debug('Loaded schema with id of "%s"', schema.id);
-      let content = JSON.stringify(schema, undefined, 4);
-      if (!content) {
-        throw new Error('Schema %s has invalid content!', name);
-      }
-      schemas[jsonName] = content;
-    } catch (err) {
-      // walk swallows errors, so we must raise them ourselves
-      if (!walkErr) {
-        walkErr = err;
+        const schema = render(json, TASKCLUSTER_SCHEMA_SCHEME, this.cfg.serviceName, this.cfg.constants);
+
+        if (schema.id || schema.$id) {
+          debug('Schema incorrectly attempts to set own id: %s', name);
+          throw new Error('Schema ' + path.join(root, name) + ' attempts to set own id!');
+        }
+
+        // We use a `taskcluster:` scheme until rootUrl is known later
+        schema.$id = libUrls.schema(TASKCLUSTER_SCHEMA_SCHEME, this.cfg.serviceName, jsonName + '#');
+        this.ajv.addSchema(schema);
```

Comment:
It looks like this gets added with the abstract schema ID, rather than the absolute ID, which (I think) would cause problems when ajv is used by the validator later.

Reply:
Ah, I see that's confirmed in the tests, and in tc-lib-api..